visualization/SensorVisualization/Sailboat.pde

```diff
@@ -0,0 +1,207 @@
+/*
+ * Sailboat
+ *
+ * (c) 2016 Brian Dolan, RoboSail
+ *
+ * This class represents a 3D model of a sailboat, including a "floating" arrow to represent
+ * wind direction.  It contains a function to draw it with a particular roll, pitch, and yaw.
+ */
+import processing.opengl.*;
+
+public class Sailboat
+{
+  private PVector location;
+  private float windAngleDeg;
+  private Arrow windDirection;
+  private float rudderAngleDeg;
+  // this should always be a positive number - the sail angle set by the client code
+  // should be between 0 and 90 degrees, as the client only really knows how far the 
+  // the sheet has been let out - within this class, we will calculate the "real" 
+  // sail angle, based on the wind direction
+  private float sailAngleLimitDeg;
+  private float rollDeg;
+  private float pitchDeg;
+  private float yawDeg;
+  
+  Sailboat(PVector location)
+  {
+    this.location = location;
+    windDirection = new Arrow(
+      new PVector(0, 0, 50), // will be replaced with value from sensor
+      color(128, 255, 128));
+  }
+  
+  // draws a sailboat model with origin at the center of the hull
+  void drawBasic()
+  {
+    pushStyle();
+    noStroke();
+    pushMatrix();
+    scale(4, 4, 4);
+    drawHull();
+    drawSail();
+    drawRudder();
+    popMatrix();
+    popStyle();
+  }
+  
+  void drawSail()
+  {
+    float sailHeight = 30;
+    float sailAngleDeg = getRealSailAngle();
+    TruncCone truncCone = new TruncCone();
+
+    pushStyle();
+    if (sailAngleDeg == windAngleDeg)
+    {
+      // the sail is luffing - tint red
+      fill(color(255, 180, 180));
```

Comment:
This is a pretty interesting idea. I suppose on the boat side we should never see the sail set to a luff position (at least with the 6 week program as we've taught it), but it's interesting to consider the algorithm for detecting when it's luffing (open loop).